### PR TITLE
Histogram backport to Metrics 3.0/4.0

### LINKED ISSUE #25519: Implement Histogram Prometheus type into MicroProfile Metrics features `mpMetrics-3.0` and `mpMetrics-4.0`
The design for the backport of the histogram functionality into mpMetrics-3.0 and mpMetrics-4.0 is identical to the mpMetrics-5.0/5.1 implementation. The Metrics 5.0/5.1 features are based on the Micrometer implementation which is the primary difference to our 3.0/4.0 implementation. We're using the same code from smallRye Metrics that Metrics v5.1 uses to parse and apply user configuration. Since the mpMetrics 3/4 features aren't based on Micrometer, we had to add our own implementation for histogram buckets into the liberty code.

UFO: Link to Metrics v5.1 UFO: https://ibm.ent.box.com/file/1268791012344?s=5qkwx6eyl2tzkmruq9hcg5i9nl0fdajv
FTS: Link to Feature Test Summary GH Issue : https://github.com/OpenLiberty/open-liberty/issues/28422
Beta Blog: Link to Beta Blog Post GH Issue : https://github.com/OpenLiberty/open-liberty/issues/28355
GA Blog: Link to GA Blog Post GH Issue : https://github.com/OpenLiberty/open-liberty/issues/28982

--------------------------------

With the existing MicroProfile Metrics specification, the technology provides a `Histogram` metric.
When exposed on the `/metrics` endpoint in Prometheus format, we expect to see an output similar to the following:
```
# TYPE application_file_sizes_mean_bytes gauge
application_file_sizes_mean_bytes 4738.231
# TYPE application_file_sizes_max_bytes gauge
application_file_sizes_max_bytes 31716
# TYPE application_file_sizes_min_bytes gauge
application_file_sizes_min_bytes 180
# TYPE application_file_sizes_stddev_bytes gauge
application_file_sizes_stddev_bytes 1054.7343037063602
# TYPE application_file_sizes_bytes summary
# HELP application_file_sizes_bytes Users file size
application_file_sizes_bytes_count 2037
application_file_sizes_bytes_sum 48123
application_file_sizes_bytes{quantile="0.5"} 4201
application_file_sizes_bytes{quantile="0.75"} 6175
application_file_sizes_bytes{quantile="0.95"} 13560
application_file_sizes_bytes{quantile="0.98"} 29643
application_file_sizes_bytes{quantile="0.99"} 31716
application_file_sizes_bytes{quantile="0.999"} 31716
```
We are primarily interested  `summary` type which includes quantiles of a specific values.

It would be of  useful if we can enhance the `mpMetrics-4.0` and `mpMetrics-5.0` features to output _histogram_ buckets alongside the quantiles/percentile output. This as a result will change the `summary` type to a `histogram` type. This functionality will be enabled through MP Config properties. This is chiefly a backport of the functionality slated to be incorporated with MicroProfile Metrics 5.1. See https://github.com/eclipse/microprofile-metrics/issues?q=is%3Aopen+is%3Aissue+milestone%3A5.1

However, for now,  the central piece would be to implement the mechanism for providing histogram bucket support in the (Liberty) implementation of MicroProfile Metrics ( of mpMetrics-3.0 and 4.0).

### PR DESCRIPTION
**Pull request by tonyreigns**
This PR introduces the backport of Histograms to Metrics 3/4 and will be beta fenced before being merged.

### CODE REVIEW

**Comment by tonyreigns** [comment]
#build

File: dev/io.openliberty.microprofile.metrics.internal.3.0/bnd.bnd [hunk 1]
```diff
@@ -44,7 +44,9 @@ Export-Package: \
 	com.ibm.ws.microprofile.metrics.helper,\
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/bnd.bnd:44]: copyright top of page

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/BucketManager.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        // Not running beta edition, throw exception
+        if (!ProductInfo.getBetaEdition()) {
+            if (!issuedBetaWarning) {
+                Tr.warning(tc, "This method is beta and is not available.");
+                issuedBetaWarning = true;
+            }
+        } else {
+            // Running beta exception, issue message if we haven't already issued one for this class
+            if (!issuedBetaMessage) {
+                Tr.info(tc, "BETA: A beta method has been invoked for the class " + this.getClass().getName() + " for the first time.");
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/BucketManager.java:158]: Not sure what the rules were, any messages not debug needs a code and translation?

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/BucketManager.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            allBuckets.put(metricName, buckets);
+        }
+
+    }
+
+    private boolean betaFenceCheck() throws UnsupportedOperationException {
+        // Not running beta edition, throw exception
+        if (!ProductInfo.getBetaEdition()) {
+            if (!issuedBetaWarning) {
+                Tr.warning(tc, "This method is beta and is not available.");
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/BucketManager.java:152]: Not sure what the rules were, any messages not debug needs a code and translation?

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/BucketManager.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    }
+
+    /**
+     * Calculates the unit String suffix and conversion factor used for later calculations
+     *
+     * @param unit String that encompasses the unit needed to calculate appropriate conversion factor and value to append
+     * @return Map.Entry<String, Double> that contains the unit string suffix and conversion factor
+     */
+    protected Map.Entry<String, Double> resolveConversionFactorXappendUnitEntry(String unit) {
+
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/BucketManager.java]: Can't remember if its' exactly the same, but it's used elsewhere. For GA we can refactor it and place it in a common place to use.

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/PrometheusBuilder30.java [hunk 5]
```diff
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- *
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/helper/PrometheusBuilder30.java:7]: copyright

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/impl/Histogram30Impl.java [hunk 6]
```diff
@@ -5,7 +5,7 @@
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
- *
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/impl/Histogram30Impl.java:8]: copyright

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/impl/Histogram30Impl.java [hunk 7]
```diff
@@ -88,12 +106,48 @@ public long getCount() {
...
     public Snapshot getSnapshot() {
+
         return reservoir.getSnapshot();
     }
 
+    public Map<String, Map<Double, BucketValue>> getBuckets() {
+        return manager.getBuckets();
+    }
+
+    public BucketManager getManager() {
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/impl/Histogram30Impl.java:117]: add javadocs (for ga)

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/impl/Timer30Impl.java [hunk 8]
```diff
@@ -5,7 +5,7 @@
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
- *
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/internal/impl/Timer30Impl.java:8]: copyright

File: dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/setup/config/DefaultBucketConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
+package io.openliberty.microprofile.metrics30.setup.config;
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/setup/config/DefaultBucketConfiguration.java:1]: Need to check, what type of copyright we put on it. We still need copy right header i think. We've wrote this for small rye, but not sure if we use the SR copyright (it's in a separate file in the repo).

Comment (Channyboy) [reply dev/io.openliberty.microprofile.metrics.internal.3.0/src/io/openliberty/microprofile/metrics30/setup/config/DefaultBucketConfiguration.java:1]: Applies to all the classes here in the package.

File: dev/io.openliberty.microprofile.metrics.internal.3.0_fat/build.gradle [hunk 10]
```diff
@@ -0,0 +1,95 @@
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0_fat/build.gradle:2]: Don't think the contents of this file needed to be copied over. This was testing something else.

File: dev/io.openliberty.microprofile.metrics.internal.3.0_fat/fat/src/io/openliberty/microprofile/metrics/internal/fat/BadHistogramTimerConfigTest.java [hunk 11]
```diff
@@ -0,0 +1,443 @@
+/**
```
Comment (Channyboy) [inline dev/io.openliberty.microprofile.metrics.internal.3.0_fat/fat/src/io/openliberty/microprofile/metrics/internal/fat/BadHistogramTimerConfigTest.java:10]: not sure if we need to append the ibm copy right too, need to ask. Applies for all the over tck classes.

**Comment by tonyreigns** [comment]
#build

**Comment by tonyreigns** [comment]
#build

**Comment by tonyreigns** [comment]
#libby

**Review by Channyboy — APPROVED** [review]
lgtm

**Pull request merged**